tests/console/snapper_cleanup.pm

```diff
@@ -73,17 +73,53 @@ sub run {
     }
 
+    my ($n_scratch_prepost, $scratchfile_mb,          $safety_margin_mb,   $initially_free);
+    my ($nmax_snapshots,    $number_limit_upper_orig, $number_limit_upper, $n);
+
+    # three test parameters; hardcoded values found appropriate in past tests
+    $n_scratch_prepost = 3;
+    $scratchfile_mb    = 1024;
+    # Expected to keep the test from filling the last 300 MB of the filesystem, poo#36838
+    $safety_margin_mb = 300 + $scratchfile_mb;
+
     assert_script_run("snapper get-config");    # get initial cfg
+    $number_limit_upper_orig = script_output("snapper get-config | awk '/^NUMBER_LIMIT\\>/{ print \$3 }'");
     assert_script_run("snapper list");          # get initial list of snap's
-    assert_script_run("snapper set-config NUMBER_MIN_AGE=0");
+
+    # check amount of free fs disk space and adapt to it
+    $initially_free = get_space("$btrfs_fs_usage | awk -F ' ' '/Free .estimated.:.*min:/{print\$3}'");    # bytes
+    $nmax_snapshots = int(($initially_free / 1_048_576 - $safety_margin_mb) / $scratchfile_mb) * 2;
+
+    # NUMBER_LIMIT setting such that disk space after cleanup
+    # allows another $n_scratch_prepost pre/post snapshot pairs
+    $number_limit_upper = $nmax_snapshots - $n_scratch_prepost * 2;
+    if ($number_limit_upper > 0) {
+        # plenty of free space? No ambition to fill it all, then.
+        $number_limit_upper = 18 if ($number_limit_upper > 18);
+        assert_script_run "snapper set-config NUMBER_LIMIT=2-$number_limit_upper NUMBER_MIN_AGE=0";
+    }
+    elsif ($number_limit_upper == 0) {
+        assert_script_run "snapper set-config NUMBER_LIMIT=0 NUMBER_MIN_AGE=0";
+    }
+    else {
+        # no cleanup test at all will be run in this case, see $n below
+        record_soft_failure('WARNING: insufficient disk space left on / for this test');
+    }
+    assert_script_run("snapper get-config");    # report customized cfg
     assert_script_run("btrfs qgroup show -pc /", 3);
     # Exclusive disk space of qgroup should be ~50% of the fs space as set with SPACE_LIMIT
     $exp_excl_space = get_space("$btrfs_fs_usage | sed -n '2p' | awk -F ' ' '{print\$3}'") / 2;
     # We need to run snapper at least couple of times to ensure it cleans up properly
-    # '4' is an arbitrary value proven by test
-    for (1 .. 4) { snapper_cleanup; }
+    # Specifically: Iteration $n - 3 to be the first to actually do clean-ups
+    $n = ($number_limit_upper >= 0) ? int(($number_limit_upper + 0.5) / ($n_scratch_prepost * 2)) + 3 : 0;
+
+    for (1 .. $n) { snapper_cleanup($n_scratch_prepost, $scratchfile_mb); }
 
-    assert_script_run("snapper set-config NUMBER_MIN_AGE=1800");
+    # tidy up and restore default settings
+    assert_script_run("snapper set-config NUMBER_LIMIT=0; snapper cleanup number; rm -fv data", 300);
+    assert_script_run("snapper set-config NUMBER_LIMIT=$number_limit_upper_orig NUMBER_MIN_AGE=1800");
```

Comment:
why 1800?

Reply:
It's default value for the snapper

Reply:
Yes, NUMBER_MIN_AGE is just reset to the known default, active at the beginning of the test. The code before my patch does likewise.

Reply:
If script_output would be stable enough, I would do either that, or same thing we do for NUMBER_LIMIT already